### DISCUSSION TITLE
vcpkg update vcpkg

### DISCUSSION THREAD

**Discussion by nurbles** [microsoft/vcpkg#22609] [Ideas]
_If this has been asked before, sorry.  I was unable to guess any search terms that located the previous question(s)._

Are there any plans to ever have **vcpkg update itself** just like it can do with any other package?

I understood why this might not have been possible with the earliest releases, but after all this time I was shocked to discover that any vcpkg update remains a completely manual process -- at least as far as the command line help and everything else I was able to find showed.

**Comment by JackBoosY** [comment]
We recommand you to use `git pull` and rerun bootstrap script to update vcpkg, but it can be considered.

**Reply by nurbles** [reply]
I was thinking that `vcpkg update vcpkg` could do something like fetch the current upgrade steps and perform them automatically (even if the procedure is somewhat different from a "normal package update" [if such a thing exists]). So it would do what you suggest and those of us that don't work with vcpkg daily wouldn't need to look up the instructions every time we want to update it (or check for an available update.)

Also, I was thinking of `apt` (on Ubuntu) and after checking discovered that I am running the current version despite only running `apt update` and `apt upgrade` periodically for _years_.  So `apt` had kept itself up-to-date along with all of my other packages and that seemed like an obvious feature that belongs in vcpkg as well.  At least, IMHO.  (smile)

Thanks for considering it.

**Reply by JackBoosY** [reply]
cc @BillyONeal  what do you think about this?

**Reply by BillyONeal** [reply]
We have considered something like this before, but given that our current use relies on vcpkg being in a git repo that users have control over, the danger of accidentally erasing a user's edits is very high so we haven't opened that can of worms. We do want a `vcpkg upgrade` command that updates your manifest when using manifest mode and separate install, but that use is itself fairly experimental at this time.

**Reply by nurbles** [reply]
Since I only use git to download something when I'm told that's the only way to get it, may I ask for an explanation?

I don't understand how an automated task that follows the exact steps that a user is told to perform manually creates a

> danger of accidentally erasing a user's edits

I would expect that the automated procedure of following the vcpkg update instructions would be _less_ likely to make a mistake than a human who is required to _manually_ type (or maybe cut'n'paste) the commands for each individual step. And along the same lines... IMHO git is broken if the act of _getting_ the current version of a package might accidentally erase a user's edits.

Sorry for my confusion.

**Reply by BillyONeal** [reply]
@nurbles Because updating vcpkg implies running `git checkout` and/or `git reset` which can erase changes if they happen while someone isn't expecting that.

**Reply by nurbles** [reply]
Interesting... neither of those git commands are mentioned in the readme about[ _installing_](https://github.com/microsoft/vcpkg#quick-start-windows) vcpkg, nor in the FAQ about [_updating_](https://github.com/microsoft/vcpkg/blob/master/docs/about/faq.md#how-do-i-update-vcpkg) vcpkg.

Why would an automated procedure use different commands than humans are instructed to use?  Or are those instructions both out of date?

**Reply by BillyONeal** [reply]
>nor in the FAQ about updating vcpkg.

That says `git pull` which (I believe) does a checkout inside. You wouldn't typically want to `git pull` in a work folder where you have lots out outstanding edits.

>Why would an automated procedure use different commands than humans are instructed to use?

People who work in git understand what git commands do and what may happen to any of their outstanding edits when they run those git commands. It's kind of like why desktop gadgets were removed in Win7: because customers didn't understand that `.gadget` was a renamed `.exe` and thought they were safe to open untrusted from the internet. It isn't that adding `.gadget` created new opportunities for attackers, it's that customers couldn't be expected to "pierce the veil" of the system's behavior.

This is why I was saying we do want such a command for manifests and there it's a lot less scary/concerning, because there we aren't mutating a git repo directly under the user's control; we would be manipulating their `vcpkg.json` and/or `vcpkg-configuration.json` which is something they should reasonably expect vcpkg to edit when they call `upgrade`.

**Reply by nurbles** [reply]
OK thanks for the explanation.  It still leaves me with a question:

Should the update instructions in the FAQ be revised to avoid the `git pull` if that command is a potential problem?  If people are following the update instructions found there, they could be causing issues?  Especially people like me who know nothing about git except for how to type the commands I'm told to type?

FWIW, I have always followed the "new install" instructions because my updates to vcpkg itself have been very infrequent and I thought it would be better to start clean than to try to upgrade -- especially by hand.

**Reply by dg0yt** [reply]
`git pull` will not simply overwrite your local edits. From `man git-pull`:

> If any of the remote changes overlap with local uncommitted changes, the merge will be automatically canceled and the work tree untouched. It is generally best to get any local changes in working order before pulling or stash them away with git-stash(1).

**Reply by nurbles** [reply]
Ah, I missed the "overwriting local edits" part.  I was thinking about helping _users_ of vcpkg because I figured that _developers_ (i.e. people who actually _have_ local edits) are intimately familiar with the update/upgrade processes and probably would not use an automated update, unless they were testing that feature itself.
